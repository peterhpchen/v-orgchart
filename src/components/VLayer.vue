<template>
  <table class="v-layer">
    <tr>
      <td>
        <v-node
          :data="data"
        >
          <template
            slot-scope="_"
          >
            <slot
              :data="_.data"
            />
          </template>
        </v-node>
      </td>
    </tr>
    <template
      v-if="isHaveChild"
    >
      <tr
        class="lines"
      >
        <td>
          <div class="downLine"/>
        </td>
      </tr>
      <tr
        class="lines"
      >
        <td
          v-for="n in data.children.length * 2"
          :key="n"
          :class="line(n)"
        />
      </tr>
      <tr
        class="nodes"
      >
        <td
          v-for="(node, index) in data.children"
          :key="key(node, index)"
        >
          <v-layer
            :data="node"
          >
            <template
              slot-scope="_"
            >
              <slot
                :data="_.data"
              />
            </template>
          </v-layer>
        </td>
      </tr>
    </template>
  </table>
</template>

<script>
import VNode from './VNode';

export default {
  name: 'VLayer',
  components: {
    VNode,
  },
  props: {
    data: {
      type: Object,
      required: true,
    },
  },
  computed: {
    isHaveChild() {
      return this.data.children && this.data.children.length > 0;
    },
  },
  methods: {
    key(node, index) {
      return node.id || index;
    },
    line(number) {
      return {
        rightLine: !this.isEven(number),
        leftLine: this.isEven(number),
      };
    },
    isEven(number){
      return number % 2 === 0;
    }
  },
};
</script>

<style lang="postcss" scoped>

</style>

