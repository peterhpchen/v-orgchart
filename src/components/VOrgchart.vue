<template>
  <div class="v-orgchart">
    <v-layer
      :data="data"
    >
      <template
        slot-scope="_"
      >
        <slot
          :data="_.data"
        >
          <div class="name">{{ _.data.name }}</div>
        </slot>
      </template>
    </v-layer>
  </div>
</template>

<script>
import VLayer from './VLayer';

export default {
  name: 'v-orgchart',
  components: {
    VLayer,
  },
  props: {
    data: {
      type: Object,
      required: true,
    },
  },
  methods: {
    key(node, index) {
      return node.id || index;
    },
  },
};
</script>

<style lang="postcss" scoped>

</style>
