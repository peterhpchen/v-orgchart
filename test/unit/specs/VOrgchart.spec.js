import { shallow } from '@vue/test-utils';
import VOrgchart from '@/components/VOrgchart';
import VLayer from '@/components/VLayer';

describe('VOrgchart.vue', () => {
  it('has VLayer', () => {
    const node = {
      name: 'A',
    };

    const wrapper = shallow(VOrgchart, {
      propsData: {
        data: node,
      },
    });

    expect(wrapper.contains(VLayer)).toBeTruthy();
  });

  it('matches snapshot', () => {
    const node = {
      name: 'A',
      children: [
        {
          name: 'B',
        },
        {
          name: 'C',
        },
        {
          name: 'D',
        },
      ],
    };

    const wrapper = shallow(VOrgchart, {
      propsData: {
        data: node,
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
