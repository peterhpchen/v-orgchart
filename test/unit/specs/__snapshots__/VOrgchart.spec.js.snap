// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VOrgchart.vue matches snapshot 1`] = `
<div class="v-orgchart">
    <table class="v-layer">
        <tr>
            <td colspan="6">
                <div class="node">
                    <div class="title"></div>
                    <div class="content">A</div>
                </div>
            </td>
        </tr>
        <tr class="lines">
            <td colspan="6">
                <div class="downLine"></div>
            </td>
        </tr>
        <tr class="lines">
            <td class="rightLine"></td>
            <td class="topLine leftLine"></td>
            <td class="topLine rightLine"></td>
            <td class="topLine leftLine"></td>
            <td class="topLine rightLine"></td>
            <td class="leftLine"></td>
        </tr>
        <tr class="nodes">
            <td colspan="2">
                <table class="v-layer">
                    <tr>
                        <td>
                            <div class="node">
                                <div class="title"></div>
                                <div class="content">B</div>
                            </div>
                        </td>
                    </tr>
                    <!---->
                </table>
            </td>
            <td colspan="2">
                <table class="v-layer">
                    <tr>
                        <td>
                            <div class="node">
                                <div class="title"></div>
                                <div class="content">C</div>
                            </div>
                        </td>
                    </tr>
                    <!---->
                </table>
            </td>
            <td colspan="2">
                <table class="v-layer">
                    <tr>
                        <td>
                            <div class="node">
                                <div class="title"></div>
                                <div class="content">D</div>
                            </div>
                        </td>
                    </tr>
                    <!---->
                </table>
            </td>
        </tr>
    </table>
</div>
`;
